funsor/distribution.py

```diff
@@ -400,10 +439,32 @@ def __call__(self, cls, args, kwargs):
 
 @to_data.register(Distribution)
 def distribution_to_data(funsor_dist, name_to_dim=None):
-    params = [to_data(getattr(funsor_dist, param_name), name_to_dim=name_to_dim)
-              for param_name in funsor_dist._ast_fields if param_name != 'value']
-    pyro_dist = funsor_dist.dist_class(**dict(zip(funsor_dist._ast_fields[:-1], params)))
     funsor_event_shape = funsor_dist.value.output.shape
+
+    # attempt to generically infer the independent output dimensions
+    instance = funsor_dist.dist_class(**{
+        k: dummy_numeric_array(domain)
+        for k, domain in zip(funsor_dist._ast_fields, (v.output for v in funsor_dist._ast_values))
+        if k != "value"
+    }, validate_args=False)
+    event_shape = broadcast_shape(instance.event_shape, funsor_dist.value.output.shape)
+    # XXX is this final broadcast_shape necessary? should just be event_shape[...]?
+    indep_shape = broadcast_shape(instance.batch_shape, event_shape[:len(event_shape) - len(instance.event_shape)])
```

Comment:
nit: could you make this more explicit and readable by declaring whether the result is negative or nonnegative? I ask because numpy slicing semantics treats the two sets very differently.
```py
reinterpreted_batch_ndims = len(event_shape) - len(instance.event_shape)
assert reinterpreted_batch_ndims > 0  # XXX is this ever nonzero?
indep_shape = broadcast_shape(instance.batch_shape, event_shape[:reinterpreted_batch_ndims])
```